test/ClientRedoEnabledTest.js

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const chai = require('chai');
+chai.should();
+chai.use(require('chai-as-promised'));
+const expect = chai.expect;
+const RC = require('./RC');
+const {
+    Client,
+    IndexOutOfBoundsError,
+    IllegalMonitorStateError
+} = require('../.');
+
+describe('ClientRedoEnabledTest', function () {
+
+    this.timeout(10000);
+
+    let cluster;
+    let client;
+
+    beforeEach(async function () {
+        cluster = await RC.createCluster(null, null);
+        await RC.startMember(cluster.id);
+    });
+
+    afterEach(async function () {
+        await client.shutdown();
+        return RC.terminateCluster(cluster.id);
+    });
+
+    it('List.get should throw on out of bounds index access', async function () {
```

Comment:
Since we consider `redoOperation` only when we receive `TargetDisconnectedError`, what is the purpose of this and `Map.unlock should throw when not locked`?

Reply:
These two tests were borrowed from https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/test/java/com/hazelcast/client/ClientRedoEnabledTest.java

And that test class was added to fix the following bug: https://github.com/hazelcast/hazelcast/issues/15157

So, I decided to keep them around to avoid running into the same bug in Node.js client.